### PR TITLE
Document limitations of Path.contains_point() and clarify its semantics

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

For reference: Issues that are concerned with this behavior: #9704, #17279, #18879.

### CODE REVIEW

File: lib/matplotlib/path.py [hunk 1]
```diff
@@ -496,7 +496,11 @@ def transformed(self, transform):
 
     def contains_point(self, point, transform=None, radius=0.0):
         """
-        Return whether the (closed) path contains the given point.
+        Return whether the area enclosed by the path contains the given points.
```
Comment (brunobeltran) [inline lib/matplotlib/path.py:499]:
```suggestion
        Return whether the area enclosed by the path contains the given point.
```

File: lib/matplotlib/path.py [hunk 2]
```diff
@@ -517,6 +521,17 @@ def contains_point(self, point, transform=None, radius=0.0):
         Returns
         -------
         bool
+
+        Notes
+        -----
+        The current algorithm has some limitations:
+
+        - The result is undefined for points exactly at the boundary
+          (including *radius*).
```
Comment (brunobeltran) [inline lib/matplotlib/path.py:530]: This makes it sound like if you make radius larger, then the area which is not defined increases, which is not true. It's just wherever the boundary ends up being the result is undefined there, unless I'm mistaken?

File: lib/matplotlib/path.py [hunk 3]
```diff
@@ -531,7 +546,11 @@ def contains_point(self, point, transform=None, radius=0.0):
 
     def contains_points(self, points, transform=None, radius=0.0):
         """
-        Return whether the (closed) path contains the given point.
+        Return whether the area enclosed by the path contains the given point.
```
Comment (brunobeltran) [inline lib/matplotlib/path.py:549]:
```suggestion
        Return whether the area enclosed by the path contains the given points.
```

File: lib/matplotlib/path.py [hunk 4]
```diff
@@ -552,6 +571,17 @@ def contains_points(self, points, transform=None, radius=0.0):
...
+        Notes
+        -----
+        The current algorithm has some limitations:
+
+        - The result is undefined for points exactly at the boundary
+          (including *radius*).
+        - The result is undefined if there is no enclosed area, i.e. all
+          vertices are on a straight line.
+        - If bounding lines start to cross each other due to *radius* shift,
+          the result is not guaranteed to be correct.
```
Comment (brunobeltran) [inline lib/matplotlib/path.py:584]: Should this section just read, "this function has various limitations, for details see `contains_point`"? I know we bend over backward to repeat keyword args and so forth in every docstring for plotting functions, but this is more of a "power-user"-facing function than a user-facing/plotting function, so I'm personally comfortable allowing the redirection (and we have historically used such redirection elsewhere in the codebase, I think?)

**Review by brunobeltran — CHANGES_REQUESTED** [review]
Small changes. No strong opinion either way on whether we should repeat the whole notes section or just use a "see also" approach.

File: lib/matplotlib/path.py (hunk 4 above)
Comment (tacaswell) [reply lib/matplotlib/path.py:584]: I think either way is fine here.  Just because power-users _can_ chase references, does not mean we should make them.

Comment (timhoffm) [reply lib/matplotlib/path.py]: No strong opinion. It's unlikely to change (little danger of becoming invalid) and explicit is slightly better than an indirection. So for now (personal preference), I'd leave this explicit.

File: lib/matplotlib/path.py (hunk 2 above)
Comment (brunobeltran) [reply lib/matplotlib/path.py]: I'd say the new version is clear now.

**Review by brunobeltran — APPROVED** [review]
Thanks for taking care of this.

**Pull request merged**